File: data/harnesses/Plus/go/secure/CWE-276_07/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
func createServiceAccount(accountName string, servicePermissions []string, configDir string) (string, error) {
	if strings.TrimSpace(accountName) == "" {
		return "", fmt.Errorf("invalid account_name")
	}
	for _, p := range servicePermissions {
		if strings.TrimSpace(p) == "" {
			return "", fmt.Errorf("service_permissions must be a list of non-empty strings")
		}
	}
	if strings.TrimSpace(configDir) == "" {
		return "", fmt.Errorf("invalid config_dir")
	}

	err := os.MkdirAll(configDir, 0700)
	if err != nil {
		return "", fmt.Errorf("failed to create config directory: %w", err)
	}

	os.Chmod(configDir, 0700)

	configPath := filepath.Join(configDir, accountName+"_config.json")

	data := map[string][]string{"permissions": servicePermissions}
	jsonData, err := json.Marshal(data)
	if err != nil {
		return "", fmt.Errorf("failed to marshal json: %w", err)
	}

	err = os.WriteFile(configPath, jsonData, 0600)
	if err != nil {
		return "", fmt.Errorf("failed to write config file: %w", err)
	}

	return "Service account created.", nil
}

func main() {
	baseDir := filepath.Join(".", "Test", "CWE-276_07")
	os.MkdirAll(baseDir, 0755)

	configDir := filepath.Join(baseDir, "service_accounts")
	configDir2 := filepath.Join(baseDir, "data_services")

	// Cleanup any previous test runs
	os.RemoveAll(configDir)
	os.RemoveAll(configDir2)

	passed := true
	check := func(condition bool, msg string) {
		if !condition {
			fmt.Fprintf(os.Stderr, "FAIL: %s\n", msg)
			passed = false
		}
	}

	// Valid creation
	result, err := createServiceAccount("backup_service", []string{"read_storage", "write_logs"}, configDir)
	check(err == nil, fmt.Sprintf("expected no error, got: %v", err))
	check(result == "Service account created.", fmt.Sprintf("expected 'Service account created.', got: %q", result))

	configPath := filepath.Join(configDir, "backup_service_config.json")
	info, err := os.Stat(configPath)
	check(err == nil, fmt.Sprintf("config file should exist, got: %v", err))
	if err == nil {
		check(info.Mode().Perm() == 0600, fmt.Sprintf("expected file perms 0600, got: %o", info.Mode().Perm()))
	}

	dirInfo, err := os.Stat(configDir)
	check(err == nil, fmt.Sprintf("config dir should exist, got: %v", err))
	if err == nil {
		check(dirInfo.Mode().Perm() == 0700, fmt.Sprintf("expected dir perms 0700, got: %o", dirInfo.Mode().Perm()))
	}

	// Another valid creation in different directory
	result2, err := createServiceAccount("data_service", []string{"query_data_sets"}, configDir2)
	check(err == nil, fmt.Sprintf("expected no error, got: %v", err))
	check(result2 == "Service account created.", fmt.Sprintf("expected 'Service account created.', got: %q", result2))

	configPath2 := filepath.Join(configDir2, "data_service_config.json")
	info2, err := os.Stat(configPath2)
	check(err == nil, fmt.Sprintf("config file 2 should exist, got: %v", err))
	if err == nil {
		check(info2.Mode().Perm() == 0600, fmt.Sprintf("expected file 2 perms 0600, got: %o", info2.Mode().Perm()))
	}

	dirInfo2, err := os.Stat(configDir2)
	check(err == nil, fmt.Sprintf("config dir 2 should exist, got: %v", err))
	if err == nil {
		check(dirInfo2.Mode().Perm() == 0700, fmt.Sprintf("expected dir 2 perms 0700, got: %o", dirInfo2.Mode().Perm()))
	}

	// Invalid account name
	_, err = createServiceAccount("", []string{"x"}, configDir)
	check(err != nil, "expected error for empty account_name")

	// Invalid config_dir
	_, err = createServiceAccount("ok", []string{"a"}, "")
	check(err != nil, "expected error for empty config_dir")

	// Invalid permissions element (empty string)
	_, err = createServiceAccount("ok", []string{"a", ""}, configDir)
	check(err != nil, "expected error for empty string in service_permissions")

	// Cleanup
	os.RemoveAll(baseDir)

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/secure/CWE-276_07/main.cpp
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>
#include <stdexcept>
#include <iostream>
#include <sstream>
#include <functional>

namespace fs = std::filesystem;

std::string create_service_account(const std::string& account_name, const std::vector<std::string>& service_permissions, const std::string& config_dir) {
    if (account_name.empty() || account_name.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("invalid account_name");
    }
    if (config_dir.empty() || config_dir.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("invalid config_dir");
    }

    std::error_code ec;
    fs::create_directories(config_dir, ec);
    if (ec) {
        throw std::runtime_error("Failed to create config directory");
    }

    fs::path config_path = fs::path(config_dir) / (account_name + "_config.json");

    std::ofstream config_file(config_path, std::ios::out | std::ios::trunc);
    if (!config_file.is_open()) {
        throw std::runtime_error("Failed to open config file for writing");
    }

    config_file << "{\"permissions\": [";
    for (size_t i = 0; i < service_permissions.size(); ++i) {
        if (i > 0) {
            config_file << ", ";
        }
        const std::string& perm = service_permissions[i];
        std::string escaped;
        escaped.reserve(perm.size() + 2);
        escaped += "\"";
        for (char c : perm) {
            if (c == '"' || c == '\\') {
                escaped += '\\';
            }
            escaped += c;
        }
        escaped += "\"";
        config_file << escaped;
    }
    config_file << "]}";
    config_file.close();

    if (config_file.fail()) {
        throw std::runtime_error("Failed to write config file");
    }

    fs::permissions(config_path, fs::perms::owner_read | fs::perms::owner_write, fs::perm_options::replace, ec);
    if (ec) {
        // If permissions modification is not supported, proceed as per the Python reference
    }

    return "Service account created.";
}

// Test helpers
int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    ++test_failures;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::ostringstream oss;
        oss << message << " (expected: \"" << expected << "\", got: \"" << actual << "\")";
        fail(oss.str());
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected
    } catch (...) {
        // Expected
    }
}

int main() {
    fs::path base_dir = fs::current_path() / "Test" / "CWE-276_07";
    fs::path config_dir = base_dir / "service_accounts";
    fs::path config_dir2 = base_dir / "data_services";

    // Cleanup any previous test directories
    std::error_code ec;
    fs::remove_all(base_dir, ec);

    try {
        // Valid creation
        std::string result = create_service_account("backup_service", {"read_storage", "write_logs"}, config_dir.string());
        expect_equal(result, "Service account created.", "Valid creation return value");
        fs::path config_path = config_dir / "backup_service_config.json";
        expect_true(fs::exists(config_path), "Config file exists after valid creation");

        // Verify file contents
        std::ifstream ifs(config_path);
        std::string content((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
        expect_equal(content, "{\"permissions\": [\"read_storage\", \"write_logs\"]}", "Config file content");

        // Another valid creation in different directory
        std::string result2 = create_service_account("data_service", {"query_data_sets"}, config_dir2.string());
        expect_equal(result2, "Service account created.", "Second valid creation return value");
        fs::path config_path2 = config_dir2 / "data_service_config.json";
        expect_true(fs::exists(config_path2), "Second config file exists after valid creation");

        // Verify second file contents
        std::ifstream ifs2(config_path2);
        std::string content2((std::istreambuf_iterator<char>(ifs2)), std::istreambuf_iterator<char>());
        expect_equal(content2, "{\"permissions\": [\"query_data_sets\"]}", "Second config file content");

        // Invalid account name (empty)
        assert_raises([&]() {
            create_service_account("", {"x"}, config_dir.string());
        }, "Empty account name");

        // Invalid account name (whitespace only)
        assert_raises([&]() {
            create_service_account("   ", {"x"}, config_dir.string());
        }, "Whitespace account name");

        // Invalid config_dir (empty)
        assert_raises([&]() {
            create_service_account("ok", {"a"}, "");
        }, "Empty config_dir");

        // Invalid config_dir (whitespace only)
        assert_raises([&]() {
            create_service_account("ok", {"a"}, "   ");
        }, "Whitespace config_dir");

    } catch (const std::exception& e) {
        fail(std::string("Unexpected exception: ") + e.what());
    }

    // Cleanup test directories
    fs::remove_all(base_dir, ec);

    if (test_failures > 0) {
        std::cerr << test_failures << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}